### DISCUSSION TITLE
Use DefaultAzureCredential for authenticating with Azure

### DISCUSSION THREAD

**Discussion by TobbenTM** [grafana/grafana#53645] [Data Sources]
Currently the Azure Monitor data source (maybe along with others) only support using application credentials or managed identity.

The latter option is by far the most ideal option for dealing with security between resources, but has one major flaw; it does not support environments where the credentials are injected any other way than the good old Azure VM kinda way.

We have the scenario where we are hosting Grafana instances in Kubernetes, and to limit the credential access of each pod in our clusters, we give the pod a service account connected to a [Azure Workload Identity](https://azure.github.io/azure-workload-identity/docs/introduction.html) issuer. This means our Grafana pods will have environment variables injected with the correct managed identity and a token, but the current implementation does not pick up on that!

I'd very much like to see some changes in this file: https://github.com/kostrse/grafana/blob/main/pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go#L78-L88

Where the type `DefaultAzureCredential` should be used over the current `ManagedAzureCredential`, they should be equivalent in the end, as seen in the official documentation: https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity#readme-defaultazurecredential

**Comment by dbg-raghulkrishna** [comment]
any update on this?

**Reply by andresmgot** [reply]
FTR, more details at https://github.com/grafana/grafana/issues/59512

**Reply by dbg-raghulkrishna** [reply]
@andresmgot  is this fixed ? the url points to the issue i created

**Reply by andresmgot** [reply]
I don't think this is part of the roadmap yet. Maybe @aangelisc have more information.

**Reply by aangelisc** [reply]
Thanks for the ping @andresmgot. Apologies @dbg-raghulkrishna but this isn't in our roadmap yet. We're planning on reviewing the Azure authentication at some point in the near future so we can build this in too 😊

**Reply by djsly** [reply]
where can we track ? we moved to Workload Identity and would like to ensure that we can configure our Datasources to use workload Identity to access the external azure specific datasources.

**Reply by ThomasEhling** [reply]
Hello, here is message to follow-up: we'd like to move to Azure Workload Identity too and are blocked by this issue.
Is there a way to track the progress on your implementation ?

**Reply by aangelisc** [reply]
Hi all, thank you for keeping this discussion active. This still isn't on our roadmap as of yet due to other FR's taking priority but I will bring this to the attention of the team to see if we can add it at some point in the future.

**Reply by patst** [reply]
Would be great to have it. The workaround described below (using the proxy sidecar) is not recommended for production use as per Azure docs and without this using Managed Identity authentication is not possible from Kubernetes clusters

**Comment by tete17** [comment]
Hey guys for anyone still facing this issue there is pretty nice workaround until the grafana team (or volunteer) gets around to changing the code.

Workload identity has compatibility mode with a proxy sidecar that translates the IMDS requests grafana does to the OAuth2 calls workload identity is based on.

They claim is not a "supported" setup for production. I am guessing they claim that projects don't use it as an excuse not to migrate to the new authentication system. All I can say is it work like a charm. All you do is [follow the guide](https://learn.microsoft.com/en-us/azure/aks/workload-identity-migrate-from-pod-identity#create-a-managed-identity) to create the appropiate resources in azure and then define the appropiate label and annotations of the pods.

I leave my setup as an example:

I use helm to deploy grafana (I suspect many other do too), but feel free to adapt it to your system. These are the additions to my values.yaml file
```yaml
# Needed for grafana to enable the managed identity feature
grafana.ini:
  azure:
      managed_identity_enabled: true

# This tells workload identity which client-id to use
serviceAccount:
  annotations:
    azure.workload.identity/client-id: "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"

# This instructs workload identity to inject the compatibility proxy sidecar 
podAnnotations:
  azure.workload.identity/inject-proxy-sidecar: "true"

# This enabled the workload identity feature on the pods
podLabels:
  azure.workload.identity/use: "true"
```

And these are my resources in terraform, all nice and tidy :^)
```hcl
data "azurerm_subscription" "current" {
}

resource "azurerm_user_assigned_identity" "grafana_azure_monitor" {
  location            = azurerm_resource_group.kubernetes.location
  name                = "Grafana-Azure-Monitor"
  resource_group_name = azurerm_resource_group.kubernetes.name
}

# https://grafana.com/docs/grafana/latest/datasources/azure-monitor/#configure-azure-active-directory-ad-authentication
resource "azurerm_role_assignment" "grafana_azure_monitor" {
  scope                = data.azurerm_subscription.current.id
  role_definition_name = "Reader"
  principal_id         = azurerm_user_assigned_identity.grafana_azure_monitor.principal_id
}

// Required for workload identity
resource "azurerm_federated_identity_credential" "grafana_azure_monitor" {
  name                = "Grafana-Azure-Monitor"
  resource_group_name = azurerm_resource_group.kubernetes.name
  parent_id           = azurerm_user_assigned_identity.grafana_azure_monitor.id

  audience = ["api://AzureADTokenExchange"]
  issuer   = azurerm_kubernetes_cluster.kubernetes-a.oidc_issuer_url
  // system:serviceaccount:${K8S_NAMESPACE}:${SERVICEACCOUNT_NAME}
  subject = "system:serviceaccount:grafana:grafana"
}
```

**Comment by slushysnowman** [comment]
Hey, just wondering, is this on the roadmap yet?

The way of doing this with workload identity that @tete17 specifies works, but is not recommended for production by Azure.

With aad-pod-identity EOL, it's basically impossible to get this working on AKS using the functionality Grafana provides, without resorting to the workaround that Azure doesn't recommend.

This is a pretty important thing to get sorted it seems.

**Comment by nbjohnson** [comment]
Any update on this? As others have stated azure pod identity is EOL at the end of September and replaced with workload identity. Has azure workload identity support been added to the roadmap? Not having this support will break Grafana integration with Azure AKS identities in a month

**Comment by weng5e** [comment]
Besides supporting only one managed identity, should this also supports choosing multiple managed identies within the Grafana application itself? Like I can use the workload identity to bind multiple managed identites to the Grafana pod, then I want to setup multiple Grafana data sources using those different managed identies.

## Sample Grafana UI mock up
<img width="491" alt="choose client Id MI" src="https://github.com/grafana/grafana/assets/22308724/4bc908cd-ea40-45bd-851c-ab60bcb8981f">

Then the Grafana running in a VMSS with multiple managed identities can support different data source using different manged identity.
![image](https://github.com/grafana/grafana/assets/22308724/d06f31da-8bfb-4a1d-9fc8-418a47044b81)

**Comment by aangelisc** [comment]
Hi all, just to keep everyone in the loop we are investigating implementing support for `WorkloadIdentity` before `PodIdentity` is EOL

**Comment by kostrse** [comment]
For information, you can track implementation of Workload Identity by following progress of this issue:
- #75404

**Comment by taleena** [comment]
Closing the loop: this is now available, and we have posted this service announcement: https://grafana.com/blog/2023/09/29/new-grafana-releases-to-support-transition-to-azure-ad-workload-identity-authentication-in-microsoft-azure/
